lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb

```diff
@@ -558,5 +567,24 @@ def handle_module_operation(node, operation)
     def current_visibility
       T.must(@visibility_stack.last)
     end
+
+    sig { returns(T.nilable(Entry::Class)) }
+    def singleton_klass
+      attached_class = @owner_stack.last
+      return unless attached_class
+
+      # Return the existing singleton class if available
+      owner = T.cast(
+        @index["#{attached_class.name}::<Class:#{attached_class.name}>"],
+        T.nilable(T::Array[Entry::SingletonClass]),
+      )
+      return owner.first if owner
+
+      # If not available, create the singleton class lazily
+      nesting = @stack + ["<Class:#{@stack.last}>"]
+      entry = Entry::SingletonClass.new(nesting, @file_path, attached_class.location, [], nil)
```

Comment:
Will there be a case that we need a parent class for `SingletonClass`? If not, maybe we can override its `initialize` method to remove the last parameter?

Reply:
As Ufuk said, we will need it for the next PR to handle proper linearization of singleton classes.

They always inherit from the singleton class of the parent class of the attached class (lol). This:

```ruby
class Foo < Bar
end

Foo.singleton_class.ancestors
# => [<Class:Foo>, <Class:Bar>, <Class:Object>, <Class:BasicObject>, Class, Module, Object, Kernel, BasicObject]
```